### DISCUSSION TITLE
Update to airflow-pgbouncer-exporter-2023.02.21-0.14.0 image soon?

### DISCUSSION THREAD

**Discussion by ahipp13** [apache/airflow#36870] [General]
Hello! I had a question regarding this image. Is this going to be updated soon? It has a "go" vulnerability that my org is going to be blocking in about a month, so I was wondering if this image is going to be updated to update some libraries used in it. Thanks!

**Comment by Taragolis** [comment]
I guess this images updated on-demand, @potiuk correct me if I wrong but I able to find only this [README.md](https://github.com/apache/airflow/blob/main/chart/dockerfiles/README.md)

BTW is any significant differences with [jbub/pgbouncer_exporter](https://hub.docker.com/r/jbub/pgbouncer_exporter/tags) image?

**Reply by ahipp13** [reply]
What do you mean by "updated on-demand"? Looks like both Alpine 3.14 and Go 1.19.6 have been EOL for quite a while:
![image](https://github.com/apache/airflow/assets/118911990/871ae73d-0490-433e-9d85-27f17f146aa5)
![image](https://github.com/apache/airflow/assets/118911990/d46d0b1a-4acf-40e8-8b3d-e126a42d018c)

I am going to try and first use the straight up images and see what happens. If that doesn't work then I will try to run the "build_and_push.sh" scripts with changing the versions. But this should have been addressed with the last 2 Helm Chart releases and should get addressed ASAP.

**Reply by potiuk** [reply]
You can generally modify the helm chart of yours can update it - no problem with it if you require it. But way better approach is that you contribute it back. Anyone including you @ahipp13 can create a PR and propose updating it. In fact it's best if those who want to have it updated, actually do it (and they can also test it locally if it works for them).

Since you are interested - would you please prepare such pr and see if new pgbouncer is ok @ahipp13 ?  Or maybe  - since your company has some expectation there - your company could pay someone to do it if you feel not skilled enough yourself?

**Reply by ahipp13** [reply]
If I find something that works for my instance I will certainty share it. I know Airflow well but not that well. I wanted to provide awareness to this issue so that people a lot smarter then me that know a lot more about Airflow can look at this and come up with a more robust solution. I get the whole opensource model but that doesn't mean I automatically have the solution or even know how to come up with it. Hence why I brought it up. If I knew the solution or even was close to one I would make a PR instead of posting this. So hopefully this puts light on this and people start looking at it.

**Reply by potiuk** [reply]
I ask you specifically. Do you want to contribtue to it or not? Do you - and your company- know that the "scanning" issue you discovered is affecite the security of airflow or not? if you know it does - can you make a security report according to Airflow security policy or not?

We specifcally ask the users of Airflow to report security issues when they know airflow dependencies  security issues to report them to us using responsible disclosure described in https://github.com/apache/airflow/security/policy . If you know a reproducible way how 3rd-party security issuse is affecting Airflow, we ask you to report it - including the reproducible scenario where it affects airflow users. Please - by all means do it if you have such scenario.

On the other hand if your goal is to satisfy your internal rules when random scans of random tools show that maybe some library has some issue that may impact your installation hsa some security issue that might or might not impact airflow security then we ask your company as a user to do one of two things:

a) show a reproducible way and scenario that it imnpacts Airflow users and report it via security@airflow.apache.org according to our policies;

b) contribute by following regular contribution process to bump the dependency version, if you cannot do a) but your internal processes require you to bump versions of some dependencies.

We canot do anything if your processes have such high expectations. But you can do a lot - for example contribute a PR to fix it.

Either a) or b) is good  - pick your way.

**Reply by potiuk** [reply]
And yes. I ask you specifically @ahipp13 to look at it. You (and your company) think it is a problem, So we ask you (and your company) to either narrow it down to actionable, reproducible security issue, or to make a PR to prevent your scanner report flagging the problem.

**Reply by Taragolis** [reply]
Just have a question with another perspective, do we actually need build and publish and maintain I guess our own images for `pgbouncer` and `pgbouncer_exporter`. And could we switch to other distributions? E.g.

Instead of our `pgbouncer` image in Chart use [bitnami/pgbouncer](https://hub.docker.com/r/bitnami/pgbouncer) as default
https://github.com/apache/airflow/blob/d18c01a7a2f04449262606d79b4f663bdb200414/chart/values.yaml#L114-L117

And instead of our [`pgbouncer_exporter`](https://github.com/jbub/pgbouncer_exporter) image in Chart use docker image provided by autor of this utility [`jbub/pgbouncer_exporter`](https://hub.docker.com/r/jbub/pgbouncer_exporter)
https://github.com/apache/airflow/blob/d18c01a7a2f04449262606d79b4f663bdb200414/chart/values.yaml#L118-L121

Both of them has support **linux/amd64** and **linux/arm64** platforms.

**Reply by potiuk** [reply]
We could potentially, but that's a bit of security issue, whether we want to have our users rely on 3rd-party images, and also makes our user quite a bit dependent on the 3rd-party being available. This is why we host those images in our own repository. They also contain specific configuration that is needed to make them run in Airflow - and while likely some of that configuration can be passed via environment variables - for example - I am not sure if all of that can be.

With Bitnami, it's likely that they are respected enough that this dependency (both security and availability) is good-enough though I lost a bit of faith in them in June 2022 (see this - extremely long issue https://github.com/bitnami/charts/issues/10539) where they broke our HelmChart's postgres integration - and many other helm charts that relied on them because they removed not-that-old charts from their indexes. And since then, to protect our users of released charts, I am rather cautious about depending on them.

This is why we build and publish our own images  and they can be easily updated by just bumping the version and building the images https://github.com/apache/airflow/tree/main/chart/dockerfiles - there are simple instructions how to build the new image and publish them (including the instructions as anyone can build and publish their images for testing - it's lilterally as easy as bumping version, setting `DOCKERHUB_USER` to your own user, publishing it there and switching to those images in chart to test it.

That's why I asked you @ahipp13 to try to submit it as PR - once you (or someone in your company) confirms that it works with such regenerated images, we should be able to rebuild and push the new images  to our repository and merge such PR.

**Reply by potiuk** [reply]
Those are the instructions BTW.

```
Those are images that are needed for the Helm Chart.

In each of the images you can find "build_and_push.sh" script that builds and pushes the image.

You need to be a PMC with direct push access to "apache/airflow" DockerHub registry to be able to push to the Airflow DockerHub registry.

You can set the DOCKERHUB_USER variable to push to your own DockerHub user if you want to test the image or build your own image.
```

**Reply by potiuk** [reply]
Hey @ahipp13 -> to save you some time and effort - I've rebuilt the images now - following the above and published them:

* airflow-pgbouncer-exporter-2024.01.19-0.16.0
* airflow-pgbouncer-2024.01.19-1.21.0

Can you pleae switch to the new images and test if it works for you ?

The PR to do that is https://github.com/apache/airflow/pull/36898  and we will merge it once you can confirm it works nicely. Our tests are running them and once the tests pass, the general smoke tests should be passing - but having someone who can test them would be a good thing to confirm that they do in production as well.

**Reply by ahipp13** [reply]
I will test them in my 3 environments and report back next week

**Reply by jedcunningham** [reply]
Just to +1 @potiuk, I'd be okay with bitnami, but definitely not the exporter image.

**Reply by ahipp13** [reply]
I have tested these images (airflow-pgbouncer-exporter-2024.01.19-0.16.0 and airflow-pgbouncer-2024.01.19-1.21.0) in my 3 airflow environments (Airflow version 2.6.3 helm chart version 1.10.0) and they have worked with no issues. I have also left this note in the PR

**Reply by potiuk** [reply]
Thanks!
